### ISSUE TITLE
package/tools: Force deploy flag does not work

### ISSUE THREAD

**Issue by covertbert** [aws/aws-cdk#15065]
Running `cdk deploy stack-name --require-approval never --force` does not work. Specifically its skips updating the stack with `(no changes)` outputted in the command line. I've also tried the following patterns:

- `cdk deploy stack-name --force`
- `cdk deploy --force stack-name`

### Reproduction Steps

Repo is private so cannot link but simply running one of the above commands with cdk toolkit version `1.108.0` should suffice. I can try and reproduce in a public repo if the person seeing this cannot reproduce.

### What did you expect to happen?

CDK would *never* skip updating stacks due to "no changes" in the template when the `--force` flag is passed.

### What actually happened?

CDK skips updating stacks due to "no changes" in the template when the `--force` flag is passed.

### Environment

  - **CDK CLI Version: ** 1.108.0
  - **Framework Version:** 1.108.0
  - **Node.js Version:** v12.22.0
  - **OS               :** MacOs Big Sur
  - **Language (Version):** TS 4.2.4

### Other

```
 ✅  service-name (no changes)
```

--- 

This is :bug: Bug Report

**Comment by peterwoodworth** [comment]
I can't seem to reproduce this, what changes did you make to your stack?

**Comment by covertbert** [comment]
I didn't make any changes to my stack. My understanding is that the point of the force flag is that it will at least trigger Cloudformation regardless of whether you've changed anything and then Cloudformation will determine what needs updating and what doesn't. In reality, for me at least, the CDK CLI just skips it completely and Cloudformation is never triggered.

**Comment by markuspxpx** [comment]
I agree with @jefftham , we should be able to overwrite the Cloudformation rule using --force, to literally force to deploy.

**Comment by Lockoo** [comment]
any updates on this? it's a really annoying issue that this flag doesn't actually work

**Comment by Lockoo** [comment]
> @Lockoo You only need to add `--require-approval never` as argument after `cdk deploy`. If you want more option type `cdk deploy --help` and search the option `--require-approval`

this won't trigger a deploy if there are no changes, it just doesnt need approval to deploy when there are.

**Comment by awerchniak** [comment]
I have the same use case - I want `cdk deploy` to re-deploy the stack even if there are no changes in the CFN template. My stack just simply builds a docker container and uploads it to ECR - while the Dockerfile itself doesn't change, the dependencies that it will be pulling via `pip` will be updated, so I want it to re-build.

I found https://github.com/aws/aws-cdk/issues/8493 which claims that you can do this with `cdk deploy --method direct`, but this didn't work for me (nor did `cdk deploy --force`).

**Comment by rix0rrr** [comment]
Hi everyone on this issue,

Please run the command again with `-v`:

```
cdk deploy --force -v
```

And observe the additional logged output.

* If the message says `skipping deployment (use --force to override)`, then CDK has short-circuited the deployment because it thinks there is nothing to do. It is *usually* right, but there might be cases in which it is wrong, and you can use `--force` to force it to update.
* If you see the text `No changes are to be performed on <STACK>`, then CloudFormation considers the change to have no effects, and in fact if we had told CloudFormation to execute it would have given us back an error. There is just nothing to do for CloudFormation, and CDK cannot change that fact.
* If you see neither of these messages, then let me know what messages you do see.

If you are planning to use `--force` to wipe away drift (that is: you've changed resources out-of-band, and now want to get them back into the state as declared in the template), then be aware: **this is not possible. CloudFormation does not support wiping away drift in this matter**. It exclusively operates on its own view of the world which it has built up over deployments done via CloudFormation, and it will not look at operations done in the mean time via the console or the CLI. If this is what you want to do, you need to look at the [CloudFormation Drift Detection](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/detect-drift-stack.html) feature.

**Comment by AndyRubio** [comment]
Our workaround has been to deploy the last version, then redeploy the current version. Far from ideal.

**Comment by sdasgupta-spectra** [comment]
Dealing with this same issue in 2025 and think @rix0rrr hit the nail on the head.

Seems like most of us are looking to force redeploy a stack that has drifted (in the CloudFormation Drift sense of the word) rather than one that has a `cdk diff` between the initially deployed stack and our local version.

Maybe out of scope here, but is there a canonical solution for ensuring that the actual resources under a CloudFormation stack are as specified by CDK aside from deploying earlier versions/deleting the stack before redeploying? If `--force` and `--method direct` are not the solution, what is?

**Comment by sdasgupta-spectra** [comment]
And to provide a couple answers to my own questions:

It looks like folks are manually fixing drift in the console (very undesirable), deleting/recreating stacks, or making tiny changes to the stack to cause a diff, deploying the new stack, and then reverting ([source](https://stackoverflow.com/questions/54386020/how-to-fix-a-drifted-aws-cloudformation-stack)).

And there is [some documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import-resolve-drift.html) on resolving drift, but this focuses on setting the drifted configuration as correct, not reapplying the CDK spec.

**Comment by mrgrain** [comment]
Closing this issue as it seems most of the reports in this issues are looking to force redeploy a stack that has drifted (in the CloudFormation Drift sense of the word) rather than one that has a cdk diff between the initially deployed stack and our local version. While we appreciate feedback on the maybe confusing naming of the option, we have no plans of changing such an established and long-standing option.

See Rico's comment for details: https://github.com/aws/aws-cdk/issues/15065#issuecomment-1824673357
If you are encountering behavior where `cdk deploy --force -v` is not behaving in the expected way laid out on the comment, please open a new issue with the verbose output.
